clang/test/Analysis/Checkers/CHERI/provenance-source.c

```diff
@@ -0,0 +1,159 @@
+  return *(char*)s; // expected-warning{{intcap with ambiguous provenance is used as capability}}
+}
+
+char no_prov(int d, char p) {
+  intptr_t a = (intptr_t)d;
+  intptr_t b = (intptr_t)p;
+
+  intptr_t s = a + b; // expected-warning{{Ambiguous provenance: LHS and RHS are not pointers}}
+                      // expected-warning@-1{{binary expression on capability types 'intptr_t' (aka '__intcap') and 'intptr_t'; it is not clear which should be used as the source of provenance; currently provenance is inherited from the left-hand side}}
+  return *(char*)s; // expected-warning{{NULL-derived capability is used as pointer}}
+}
+
+char right_prov_cond(int d, char *p, int x) {
+  intptr_t a = d;
+  intptr_t b;
+  if (d > 42)
+    b = (intptr_t)p;
+  else
+    b = x;
+
+  intptr_t s = a + b; // expected-warning{{Ambiguous provenance: LHS is integer, RHS is pointer}}
+                      // expected-warning@-1{{Ambiguous provenance: LHS and RHS are not pointers}}
+                      // expected-warning@-2{{binary expression on capability types 'intptr_t' (aka '__intcap') and 'intptr_t'; it is not clear which should be used as the source of provenance; currently provenance is inherited from the left-hand side}}
+  return *(char*)s;// expected-warning{{intcap with ambiguous provenance is used as capability}}
+                   // expected-warning@-1{{NULL-derived capability is used as pointer}}
+}
+
+intptr_t no_bug(int *d, int *p) {
+  intptr_t a = (intptr_t)d;
+  intptr_t b = (intptr_t)p;
+  d = p + 20;
+
+  a = b;
+  return a;
+}
+
+char add_const(int *p, int x) {
+  intptr_t a = (intptr_t)p + 42;
+  intptr_t b = (intptr_t)x;
+
+  intptr_t s =  a | b; // expected-warning{{Ambiguous provenance: LHS is pointer, RHS is integer}}
+                       // expected-warning@-1{{binary expression on capability types 'intptr_t' (aka '__intcap') and 'intptr_t'; it is not clear which should be used as the source of provenance; currently provenance is inherited from the left-hand side}}
+  return *(char*)s; // expected-warning{{intcap with ambiguous provenance is used as capability}}
+}
+
+char add_var(int *p, int x, int c) {
+  intptr_t a = (intptr_t)p + c;
+  intptr_t b = (intptr_t)x;
+
+  intptr_t s =  a | b; // expected-warning{{Ambiguous provenance: LHS is pointer, RHS is integer}}
+                      // expected-warning@-1{{binary expression on capability types 'intptr_t' (aka '__intcap') and 'intptr_t'; it is not clear which should be used as the source of provenance; currently provenance is inherited from the left-hand side}}
+  return *(char*)s; // expected-warning{{intcap with ambiguous provenance is used as capability}}
+}
+
+
+uintptr_t fn1(char *str, int f, int i) {
+    str++;
+    return ((intptr_t)str & (intptr_t)(f)); // expected-warning{{Ambiguous provenance: LHS is pointer, RHS is integer}}
```

Comment:
This should be well-defined and unambiguous (note how we do _not_ diagnose this in Clang)